File: Competitions/Aces/aces_pre_coders_v8/Gauntlet Quest from P.py
from heapq import heappush, heappop
from itertools import product


def find_distances(graph, k, n, stones):
    print("****************")
    distances = [[float('inf')]*(1 << (k)) for _ in range(n)]
    dis = 0
    node = 0
    ss = stones[0]
    heap = [(dis, node, ss)]

    target_node = n-1
    target_stone = (1 << k) - 1
    while True:
        dis, node, ss = heappop(heap)
        print(dis, node, ss)
        if (dis >= distances[node][ss]):
            continue
        distances[node][ss] = dis
        if (node == target_node and ss == target_stone):
            break
        for ch, t in graph[node]:
            new_stone = ss | stones[ch]
            new_dis = dis + t
            if (new_dis >= distances[ch][new_stone]):
                continue
            heappush(heap, (new_dis, ch, new_stone))
    print(distances[n-1])
    return distances[n-1]


def solve(distances, k):
    target = (1 << k)-1
    best = float('inf')
    for s1, d1 in enumerate(distances):
        for s2, d2 in enumerate(distances):
            stone = s1 | s2
            if (stone == target):
                best = min(best, max(d1, d2))
    return best


def solve2(distances, k):
    target = (1 << k)-1
    itt = product(enumerate(distances), enumerate(distances))
    return min((s1 | s2 != target, max(d1, d2)) for ((s1, d1), (s2, d2)) in itt)[1]


def main():
    n, m, k = map(int, input().split())
    stones = [0]*n
    graph = [[] for _ in range(n)]
    for i in range(n):
        _, *cs = map(int, input().split())
        for c in cs:
            c -= 1
            stones[i] |= 1 << c
    for _ in range(m):
        u, v, t = map(int, input().split())
        u -= 1
        v -= 1
        graph[u].append((v, t))
        graph[v].append((u, t))
    distances = find_distances(graph, k, n, stones)
    print(solve2(distances, k))


main()

File: Competitions/Aces/aces_pre_coders_v8/Gauntlet Quest from P.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = int64_t;
using vecl = vector<ll>;
using pll = pair<ll, ll>;
using vecpl = vector<pll>;
using mapl = unordered_map<ll, vecl>;
using vecvecl = vector<vecl>;
using graph_t = vector<vecpl>;

ll solve(vecl &stones, graph_t &graph, ll n, ll k)
{
    ll distances[n][1 << k];
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < (1 << k); j++)
        {
            distances[i][j] = LONG_LONG_MAX;
        }
    }

    auto heap = priority_queue<tuple<ll, ll, ll>>();

    ll target_node = n - 1;
    ll target_stone = (1 << k) - 1;
    ll dis = 0;
    ll node = 0;
    ll stone = stones[0];

    heap.push(make_tuple(-dis, node, stone));

    while (1)
    {
        tuple<ll, ll, ll> state = heap.top();

        ll dis = -get<0>(state);
        ll node = get<1>(state);
        ll stone = get<2>(state);

        // cout << dis << " " << node << " " << stone << endl;
        heap.pop();

        if (dis >= distances[node][stone])
            continue;

        distances[node][stone] = dis;

        if (node == target_node && stone == target_stone)
        {
            break;
        }

        for (auto c : graph[node])
        {
            ll new_dis = dis + c.second;
            ll new_stone = stones[c.first] | stone;

            if (new_dis >= distances[c.first][new_stone])
                continue;

            heap.push(make_tuple(-new_dis, c.first, new_stone));
        }
    }

    ll ans = LONG_LONG_MAX;
    for (int i = 0; i <= target_stone; i++)
    {
        for (int j = 0; j <= i; j++)
        {
            ll s = i | j;

            if (s != target_stone)
                continue;
            
            ans = min(ans, max(distances[target_node][i], distances[target_node][j]));
        }
    }
    return ans;
}

int main()
{
    int n, m, k;
    cin >> n >> m >> k;
    vecl stones = vecl(n);
    graph_t graph = graph_t(n);

    for (int i = 0; i < n; i++)
    {
        int nc;
        cin >> nc;
        int c;
        for (int j = 0; j < nc; j++)
        {
            cin >> c;
            c--;
            stones[i] |= (1 << c);
        }
    }

    for (int i = 0; i < m; i++)
    {
        ll t, u, v;
        cin >> u >> v >> t;
        u--;
        v--;
        graph[u].push_back(make_pair(v, t));
        graph[v].push_back(make_pair(u, t));
    }

    cout << solve(stones, graph, n, k) << endl;
    return 0;
}